module/x/gravity/keeper/grpc_query.go

```diff
@@ -1,354 +1,407 @@
+}
+
+func (k Keeper) PendingContractCallTxEthereumSignatures(c context.Context, req *types.PendingContractCallTxEthereumSignaturesRequest) (*types.PendingContractCallTxEthereumSignaturesResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	val, err := k.getSignerValidator(ctx, req.Address)
+	if err != nil {
+		return nil, err
+	}
+	var calls []*types.ContractCallTx
+	k.IterateOutgoingTxsByType(ctx, types.ContractCallTxPrefixByte, func(_ []byte, otx types.OutgoingTx) bool {
+		sig := k.GetEthereumSignature(ctx, otx.GetStoreIndex(), val)
+		if len(sig) == 0 { // it's pending
+			call, ok := otx.(*types.ContractCallTx)
+			if !ok {
+				panic(sdkerrors.Wrapf(types.ErrInvalid, "couldn't cast to contract call for %s", otx))
+			}
+			calls = append(calls, call)
+		}
+		return false
+	})
+	return &types.PendingContractCallTxEthereumSignaturesResponse{Calls: calls}, nil
+}
+
+func (k Keeper) LastSubmittedEthereumEvent(c context.Context, req *types.LastSubmittedEthereumEventRequest) (*types.LastSubmittedEthereumEventResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	res := &types.LastSubmittedEthereumEventResponse{}
+
+	valAddr, err := sdk.ValAddressFromBech32(req.Address)
+	if err != nil {
+		return nil, err
+	}
+
+	res.EventNonce = k.GetLastEventNonceByValidator(ctx, valAddr)
+
+	return &types.LastSubmittedEthereumEventResponse{}, nil
+}
+
+func (k Keeper) BatchTxFees(c context.Context, req *types.BatchTxFeesRequest) (*types.BatchTxFeesResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	res := &types.BatchTxFeesResponse{}
+
+	k.IterateOutgoingTxsByType(ctx, types.BatchTxPrefixByte, func(key []byte, otx types.OutgoingTx) bool {
+		btx, _ := otx.(*types.BatchTx)
+		for _, tx := range btx.Transactions {
+			res.Fees = append(res.Fees, tx.Erc20Fee.GravityCoin())
+		}
+		return false
+	})
+
+	return res, nil
+}
+
+func (k Keeper) ERC20ToDenom(c context.Context, req *types.ERC20ToDenomRequest) (*types.ERC20ToDenomResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	cosmosOriginated, denom := k.ERC20ToDenomLookup(ctx, req.Erc20)
+	res := &types.ERC20ToDenomResponse{
+		Denom:            denom,
+		CosmosOriginated: cosmosOriginated,
+	}
+	return res, nil
+}
+
+func (k Keeper) DenomToERC20(c context.Context, req *types.DenomToERC20Request) (*types.DenomToERC20Response, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	cosmosOriginated, erc20, err := k.DenomToERC20Lookup(ctx, req.Denom)
+	if err != nil {
+		return nil, err
+	}
+	res := &types.DenomToERC20Response{
+		Erc20:            erc20.Hex(),
+		CosmosOriginated: cosmosOriginated,
+	}
+	return res, nil
+}
+
+func (k Keeper) BatchedSendToEthereums(c context.Context, req *types.BatchedSendToEthereumsRequest) (*types.BatchedSendToEthereumsResponse, error) {
+	ctx := sdk.UnwrapSDKContext(c)
+	res := &types.BatchedSendToEthereumsResponse{}
+
+	k.IterateOutgoingTxs(ctx, types.BatchTxPrefixByte, func(_ []byte, outgoing types.OutgoingTx) bool {
```

Comment:
paginate

Reply:
Done

Reply:
This is going to be tricky. Since we have to do two layers of iteration to get the Batched Send To Ethereums. We need to iterate through the batches, and then through their internal transactions. The SDK query paginate function does accumulation only at one level. It doesn't have a mechanism to return multiple hits for a callback. 

In order to make pagination work, we would basically need to rewrite the pagination function or copy it